stack-graphs/src/stitching.rs

```diff
@@ -582,80 +732,78 @@ impl ForwardPartialPathStitcher {
     }
 
     /// Attempts to extend one partial path as part of the algorithm.  When calling this function,
-    /// you are responsible for ensuring that `db` already contains all of the possible partial
-    /// paths that we might want to extend `partial_path` with.
-    fn stitch_partial_path(
+    /// you are responsible for ensuring that `db` already contains all of the possible appendables
+    /// that we might want to extend `partial_path` with.
+    fn extend<A, Db>(
         &mut self,
         graph: &StackGraph,
         partials: &mut PartialPaths,
-        db: &mut Database,
+        db: &mut Db,
         partial_path: &PartialPath,
-        cycle_detector: AppendingCycleDetector<OwnedOrDatabasePath>,
-    ) -> usize {
-        self.candidate_partial_paths.clear();
-        if graph[partial_path.end_node].is_root() {
-            // The join node is root, so there's no need to use half-open symbol stacks here, as we
-            // do for [`PartialPath::concatenate`][].
-            let key = SymbolStackKey::from_partial_symbol_stack(
-                partials,
-                db,
-                partial_path.symbol_stack_postcondition,
-            );
-            db.find_candidate_partial_paths_from_root(
-                graph,
-                partials,
-                Some(key),
-                &mut self.candidate_partial_paths,
-            );
-        } else {
-            db.find_candidate_partial_paths_from_node(
-                graph,
-                partials,
-                partial_path.end_node,
-                &mut self.candidate_partial_paths,
-            );
+        cycle_detector: AppendingCycleDetector<H>,
+    ) -> usize
+    where
+        A: Appendable,
+        Db: Candidates<H> + ToAppendable<H, A>,
+    {
+        copious_debugging!("    Extend {}", partial_path.display(graph, partials));
+
+        // check is path is cyclic, in which case we do not extend it
+        let has_precondition_variables = partial_path.symbol_stack_precondition.has_variable()
+            || partial_path.scope_stack_precondition.has_variable();
+        let cycles = cycle_detector
+            .is_cyclic(graph, partials, db, &mut self.appended_paths)
+            .expect("cyclic test failed when stitching partial paths");
+        let cyclic = match has_precondition_variables {
+            // If the precondition has no variables, we allow cycles that strenghten the
+            // precondition, because we know they cannot strenghten the precondition of
+            // the overall path.
+            false => !cycles
+                .into_iter()
+                .all(|c| c == Cyclicity::StrengthensPrecondition),
+            // If the precondition has variables, do not allow any cycles, not even those
+            // that strengthen the precondition. This is more strict than necessary. Better
+            // might be to disallow precondition strenghtening cycles only if they would
```

Comment:
typo: strengthening (also should we file an issue about this so we don’t forget?)